main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,278 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+      case Expression.Int8(lit) => Expression.Int8(lit)
+      case Expression.Int16(lit) => Expression.Int16(lit)
+      case Expression.Int32(lit) => Expression.Int32(lit)
+      case Expression.Int64(lit) => Expression.Int64(lit)
+      case Expression.BigInt(lit) => Expression.BigInt(lit)
+      case Expression.Str(lit) => Expression.Str(lit)
+      case Expression.LoadBool(exp, offset) => Expression.LoadBool(optimize(exp), offset)
+      case Expression.LoadInt8(exp, offset) => Expression.LoadInt8(optimize(exp), offset)
+      case Expression.LoadInt16(exp, offset) => Expression.LoadInt16(optimize(exp), offset)
+      case Expression.LoadInt32(exp, offset) => Expression.LoadInt32(optimize(exp), offset)
+      case Expression.StoreBool(exp, offset, v) => Expression.StoreBool(optimize(exp), offset, optimize(v))
+      case Expression.StoreInt8(exp, offset, v) => Expression.StoreInt8(optimize(exp), offset, optimize(v))
+      case Expression.StoreInt16(exp, offset, v) => Expression.StoreInt16(optimize(exp), offset, optimize(v))
+      case Expression.StoreInt32(exp, offset, v) => Expression.StoreInt32(optimize(exp), offset, optimize(v))
+      case Expression.Var(sym, tpe, loc) => Expression.Var(sym, tpe, loc)
+      case Expression.Ref(sym, tpe, loc) => Expression.Ref(sym, tpe, loc)
+      case Expression.Lambda(args, body, tpe, loc) => Expression.Lambda(args, optimize(body), tpe, loc)
+      case Expression.Hook(hook, tpe, loc) => Expression.Hook(hook, tpe, loc)
+      case Expression.MkClosure(lambda, freeVars, tpe, loc) =>
+        Expression.MkClosure(optimize(lambda).asInstanceOf[Expression.Lambda], freeVars, tpe, loc)
+      case Expression.MkClosureRef(ref, freeVars, tpe, loc) =>
+        Expression.MkClosureRef(optimize(ref).asInstanceOf[Expression.Ref], freeVars, tpe, loc)
+      case Expression.ApplyRef(sym, args, tpe, loc) => Expression.ApplyRef(sym, optimizeExps(args), tpe, loc)
+      case Expression.ApplyTail(sym, formals, actuals, tpe, loc) =>
+        Expression.ApplyTail(sym, formals, optimizeExps(actuals), tpe, loc)
+      case Expression.ApplyHook(hook, args, tpe, loc) => Expression.ApplyHook(hook, optimizeExps(args), tpe, loc)
+      case Expression.Apply(exp, args, tpe, loc) => Expression.Apply(optimize(exp), optimizeExps(args), tpe, loc)
+      case Expression.Unary(op, exp, tpe, loc) => Expression.Unary(op, optimize(exp), tpe, loc)
+      case Expression.Binary(op, exp1, exp2, tpe, loc) => optimizeBinaryExpression(e0.asInstanceOf[Expression.Binary])
+      case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+        val e1 = optimize(exp1)
+        e1 match {
+          // Elimination of dead branches (e.g. if (true) e1 else e2).
+          case Expression.True => optimize(exp2)
+          case Expression.False => optimize(exp3)
+          case _ =>
+            // Elimination of redundant branching (e.g. if(c1, if(c2, e2, e3), e3) -> if (c1 && c2, e2, e3))
```

Comment:
Right now, elimination of redundant branching will work for something like this:

```
def f: Int =
  let x = 2;
    if (x > 1)
      (if (x > 0) 5 else 2)
    else
      5
```

but not for this:

```
def f: Int =
  let x = 2;
    if (x > 1)
      (if (x > 0) 5*x else 2)
    else
      5*x
```

This is because the `SourceLocation` parameters in the two `5*x` expression differ, and thus they are not considered the same by `==`.  I could write a custom Expression equality function that ignores the `SourceLocation` fields in Expression to address this.  Any thoughts?